auth/user_mgt.go

```diff
@@ -84,6 +86,217 @@ type ExportedUserRecord struct {
 	PasswordSalt string
 }
 
+// UserProvider represents a user identity provider.
+//
+// One or more user providers can be specified for each user when importing in bulk.
+// See UserToImport type.
+type UserProvider struct {
+	UID         string
+	ProviderID  string
+	Email       string
+	DisplayName string
+	PhotoURL    string
+}
+
+// UserToImport represents a user account that can be bulk imported into Firebase Auth.
+type UserToImport struct {
+	info   *identitytoolkit.UserInfo
+	claims map[string]interface{}
+}
+
+func (u *UserToImport) userInfo() *identitytoolkit.UserInfo {
+	if u.info == nil {
+		u.info = &identitytoolkit.UserInfo{}
+	}
+	return u.info
+}
+
+func (u *UserToImport) validatedUserInfo() (*identitytoolkit.UserInfo, error) {
+	if u.info == nil {
+		return nil, fmt.Errorf("no parameters are set on the user to import")
+	}
+	info := u.info
+	if err := validateUID(info.LocalId); err != nil {
+		return nil, err
+	}
+	if info.Email != "" {
+		if err := validateEmail(info.Email); err != nil {
+			return nil, err
+		}
+	}
+	if info.PhoneNumber != "" {
+		if err := validatePhone(info.PhoneNumber); err != nil {
+			return nil, err
+		}
+	}
+	if len(u.claims) > 0 {
+		cc, err := marshalCustomClaims(u.claims)
+		if err != nil {
+			return nil, err
+		}
+		info.CustomAttributes = cc
+	}
+
+	for _, p := range info.ProviderUserInfo {
+		if p.RawId == "" {
+			return nil, fmt.Errorf("user provdier must specify a uid")
+		}
+		if p.ProviderId == "" {
+			return nil, fmt.Errorf("user provider must specify a provider ID")
+		}
+	}
+	return info, nil
+}
+
+// UID setter. This field is required.
+func (u *UserToImport) UID(uid string) *UserToImport {
```

Comment:
Setters seem like an anti-pattern in go. It seems like you're making a fluent API, but then it's a little weird for a fluent API to mutate. Why are we providing this API rather than letting them supply a struct literal that's filled in already?

Reply:
We've used this for `UserToCreate` and `UserToUpdate` before. This enables us to handle default values of fields well (e.g. `EmailVerified`). I think in the future we can provide an unmarshal method to easily construct a user from a json string.

Reply:
Well, it's good to be consistent with UserToCreate and UserToUpdate, so SGTM.